examples/mnist/main-cnn.cpp

```diff
@@ -0,0 +1,250 @@
+#include "ggml/ggml.h"
+
+#include "common.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstring>
+#include <ctime>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <algorithm>
+
+#if defined(_MSC_VER)
+#pragma warning(disable: 4244 4267) // possible loss of data
+#endif
+
+struct mnist_model {
+    struct ggml_tensor * conv2d_1_kernel;
+    struct ggml_tensor * conv2d_1_bias;
+    struct ggml_tensor * conv2d_2_kernel;
+    struct ggml_tensor * conv2d_2_bias;
+    struct ggml_tensor * dense_weight;
+    struct ggml_tensor * dense_bias;
+    struct ggml_context * ctx;
+};
+
+static bool load_conv2d_kernel(const std::string & fname, struct ggml_tensor * kernel)
+{
+    auto fin = std::ifstream(fname, std::ios::binary);
+    if (!fin) {
+        fprintf(stderr, "%s: failed to open '%s'\n", __func__, fname.c_str());
+        return false;
+    }
+    const int ne = kernel->ne[0] * kernel->ne[1] * kernel->ne[2] * kernel->ne[3];
+    std::vector<float> kernel_f32(ne);
+    fin.read(reinterpret_cast<char *>(kernel_f32.data()), ne*sizeof(float));
+    fin.close();
+    // convert kernel to fp16
+    ggml_fp32_to_fp16_row(kernel_f32.data(), (ggml_fp16_t *)kernel->data, ne);
+    return true;
+}
+
+static bool load_conv2d_bias(const std::string & fname, struct ggml_tensor * bias)
+{
+    auto fin = std::ifstream(fname, std::ios::binary);
+    if (!fin) {
+        fprintf(stderr, "%s: failed to open '%s'\n", __func__, fname.c_str());
+        return false;
+    }
+    const int ne0 = bias->ne[0];
+    const int ne1 = bias->ne[1];
+    const int ne = bias->ne[2];
+    std::vector<float> bias_f32(ne);
+    fin.read(reinterpret_cast<char *>(bias_f32.data()), ne*sizeof(float));
+    fin.close();
+    // TODO: is there a better way to convert 1d bias to 4d bias?
+    for (int i = 0; i < ne; i++) {
+        for (int j = 0; j < ne0*ne1; j++) {
+            ggml_set_f32_1d(bias, i*ne0*ne1 + j, bias_f32[i]);
+        }
+    }
```

Comment:
We probably need to add 4d overloads for `ggml_set_...` but for now this should be fine